pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1797,55 +1798,88 @@ impl RemoteTimelineClient {
         Ok(())
     }
 
+    /// Returns true if a can bypass b, i.e. if the operations don't conflict.
     ///
-    /// Pick next tasks from the queue, and start as many of them as possible without violating
-    /// the ordering constraints.
-    ///
-    /// The caller needs to already hold the `upload_queue` lock.
-    fn launch_queued_tasks(self: &Arc<Self>, upload_queue: &mut UploadQueueInitialized) {
-        while let Some(next_op) = upload_queue.queued_operations.front() {
-            // Can we run this task now?
-            let can_run_now = match next_op {
-                UploadOp::UploadLayer(..) => {
-                    // Can always be scheduled.
-                    true
-                }
-                UploadOp::UploadMetadata { .. } => {
-                    // These can only be performed after all the preceding operations
-                    // have finished.
-                    upload_queue.inprogress_tasks.is_empty()
-                }
-                UploadOp::Delete(..) => {
-                    // Wait for preceding uploads to finish. Concurrent deletions are OK, though.
-                    upload_queue.num_inprogress_deletions == upload_queue.inprogress_tasks.len()
-                }
+    /// TODO: consider moving this and other associated logic into UploadOp and UploadQueue.
+    fn can_bypass(a: &UploadOp, b: &UploadOp) -> bool {
```

Comment:
Let's unit test this against the most important invariants:
- A layer upload must happen before an index that references it
- A layer deletion must happen after an index that de-references it
- If a layer name that is re-used, the second upload must come after an index that de-references the earlier layer of the same name
- ...whichever others we can think of